communication/state_machine.cpp

```diff
@@ -336,89 +331,75 @@ bool State_machine::update(State_machine* state_machine)
             break;
     }
 
-    // Check simulation mode
-    if ( mav_modes_is_hil(state_machine->state_.mav_mode()) == true )
-    {
-        mode_new |= MAV_MODE_FLAG_HIL_ENABLED;
-    }
-    else
-    {
-        mode_new &= ~MAV_MODE_FLAG_HIL_ENABLED;
-    }
 
     // Finally, write new modes and states
-    state_machine->state_.mav_mode_       = mode_new;
     state_machine->state_.mav_state_       = state_new;
     state_machine->state_.mav_mode_custom = mode_custom_new;
 
     // overwrite internal state of joystick
-    state_machine->manual_control_.set_mode_of_source(mode_new);
+    state_machine->manual_control_.set_mode_of_source(state_machine->state_.mav_mode_);
 
     return true;
 }
 
 
-
-
-
-bool State_machine::set_mode_guided(bool guided)
+bool State_machine::set_ctrl_mode(Mav_mode mode)
 {
-    // if already in desired state, return true
-    if(state_.is_guided() == guided)
+    bool success = true;
+
+    // check if we can set/clear stabilize flag
+    if(!is_set_stabilize_allowed(mode.is_stabilize()))
     {
-        return true;
+        print_util_dbg_print("[STATE_MACHINE]: prevented passing to stabilize because imu or ahrs not ready\r\n");
+        success = false;
     }
 
-    if(guided)
+    // check if we can set/clear guided flag
+    if(!is_set_guided_allowed(mode.is_guided()))
     {
-        // if position_estimation is not healthy, abort
-        if(!position_estimation_.healthy())
-        {
-            print_util_dbg_print("[STATE_MACHINE]: prevented passing to guided because position estimation is not healthy\r\n");
-            return false;
-        }
-
-        // set guided flag
-        state_.mav_mode_ |= MAV_MODE_FLAG_GUIDED_ENABLED;
-        print_util_dbg_print("[STATE_MACHINE]: passing to guided mode\r\n");
+        print_util_dbg_print("[STATE_MACHINE]: prevented passing to guided because position estimation is not healthy\r\n");
+        success = false;
     }
-    else
+
+    if(success)
     {
-        // clear guided flag
-        state_.mav_mode_ &= ~MAV_MODE_FLAG_GUIDED_ENABLED;
-        print_util_dbg_print("[STATE_MACHINE]: leave guided mode\r\n");
+        state_.mav_mode_.set_ctrl_mode(static_cast<Mav_mode::ctrl_mode_t>(mode.bits()));
```

Comment:
With undefined mode I mean a mode where the controller is not defined such as guided but not stabilized. However, arming in an undefined state is not a problem, since the servos are set to failsafe if the mode is not defined.